client.go

```diff
@@ -418,140 +430,120 @@ func (c *Client) Browse(req *ua.BrowseRequest) (*ua.BrowseResponse, error) {
 	return res, err
 }
 
-// todo(fs): this is not done yet since we need to be able to register
-// todo(fs): monitored items.
-type Subscription struct {
-	res *ua.CreateSubscriptionResponse
-}
-
-// todo(fs): return subscription object with channel
-func (c *Client) Subscribe(intv time.Duration) (*Subscription, error) {
+// Subscribe creates a Subscription with given parameters. Parameters that have not been set
+// (have zero values) are overwritten with default values.
+// See opcua.DefaultSubscription* constants
+func (c *Client) Subscribe(params *SubscriptionParameters) (*Subscription, error) {
+	if params == nil {
+		params = &SubscriptionParameters{}
+	}
+	params.setDefaults()
 	req := &ua.CreateSubscriptionRequest{
-		RequestedPublishingInterval: float64(intv / time.Millisecond),
-		RequestedLifetimeCount:      60,
-		RequestedMaxKeepAliveCount:  20,
+		RequestedPublishingInterval: float64(params.Interval / time.Millisecond),
+		RequestedLifetimeCount:      params.LifetimeCount,
+		RequestedMaxKeepAliveCount:  params.MaxKeepAliveCount,
 		PublishingEnabled:           true,
+		MaxNotificationsPerPublish:  params.MaxNotificationsPerPublish,
+		Priority:                    params.Priority,
 	}
 
 	var res *ua.CreateSubscriptionResponse
 	err := c.Send(req, func(v interface{}) error {
 		return safeAssign(v, &res)
 	})
-	return &Subscription{res}, err
+	if err != nil {
+		return nil, err
+	}
+	if res.ResponseHeader.ServiceResult != ua.StatusOK {
+		return nil, res.ResponseHeader.ServiceResult
+	}
+
+	sub := Subscription{
+		res.SubscriptionID,
+		res.RevisedPublishingInterval,
+		res.RevisedLifetimeCount,
+		res.RevisedMaxKeepAliveCount,
+		params.Notifs,
+		c,
+	}
+	c.subscriptions[sub.SubscriptionID] = sub
+
+	return &sub, nil
 }
 
-type PublishNotificationData struct {
-	SubscriptionID uint32
-	Error          error
-	Value          interface{}
+func (c *Client) notifySubscriptions(ctx context.Context, err error) {
+	for _, sub := range c.subscriptions {
+		go func(s Subscription) {
+			s.sendNotification(ctx, &PublishNotificationData{Error: err})
+		}(sub)
+	}
 }
 
-func (c *Client) Publish(notif chan<- PublishNotificationData) {
-	// Empty SubscriptionAcknowledgements for first PublishRequest
-	var acks = make([]*ua.SubscriptionAcknowledgement, 0)
+func (c *Client) notifySubscription(ctx context.Context, response *ua.PublishResponse) {
```

Comment:
This method is fishy since it is in the `Client` but is only called from within the `Subscription` and then only calls subscription methods. Either this should be a method of `Subscription` or there should only be one run loop. See other comment on the same topic.

Reply:
It's in the `Client`, because only the `Client` is supposed to know the subscription to which the notifications in the response belong to. Remember that a `Publish` sent from one subscription's loop can return with notifications for another subscription.

No, there shouldn't be only one loop. We should maintain at least one loop per subscription so that the server always has a Publish request to send notifications for each subscription. In the future, we should extend the API to allow for queuing of Publish requests to try and maintain a certain number of Publish requests for any given subscription (See Part 4, Chapter 5.13.5.1)

Reply:
Although, we don't start the loops. The user does. Nevertheless, he should be able to start multiple loops